examples/vision/randaugment.py

```diff
@@ -0,0 +1,296 @@
+    # Input to `augment()` is a TensorFlow tensor which
+    # is not supported by `imgaug`. This is why we first
+    # convert it to its `numpy` variant.
+    return rand_aug(images=images.numpy())
+
+
+"""
+## Create TensorFlow `Dataset` objects
+
+There's one problem, though. We cannot map our `augment()` function to a TensorFlow
+`Dataset` object. To tackle this problem, we will make use of the
+[`tf.py_function`](https://www.tensorflow.org/api_docs/python/tf/py_function) that can
+convert a Python function into a TensorFlow op.
+"""
+
+train_ds_rand = (
+    tf.data.Dataset.from_tensor_slices((x_train, y_train))
+    .shuffle(BATCH_SIZE * 100)
+    .batch(BATCH_SIZE)
+    .map(
+        lambda x, y: (tf.py_function(augment, [x], [tf.float32]), y),
+        num_parallel_calls=AUTO,
+    )
+    # The returned output contains an unncessary axis of
+    # 1-D and we need to remove it.
+    .map(lambda x, y: (tf.squeeze(x), y), num_parallel_calls=AUTO)
+    .prefetch(AUTO)
+)
+
+test_ds = (
+    tf.data.Dataset.from_tensor_slices((x_test, y_test))
+    .batch(BATCH_SIZE)
+    .prefetch(AUTO)
+)
+
+"""
+**Note about using `tf.py_function`**:
+
+As our `augment()` function is not a native TensorFlow operation chances are likely that
+it can turn into an expensive operation. This is why it is much better to apply it
+_after_ batching our dataset.
+"""
+
+"""
+For comparison purposes, let's also define a simple augmentation pipeline consisting of
+random flips, random rotations, and random zoomings.
+"""
+
+simple_aug = tf.keras.Sequential(
+    [
+        layers.experimental.preprocessing.RandomFlip("horizontal"),
+        layers.experimental.preprocessing.RandomRotation(factor=0.02),
+        layers.experimental.preprocessing.RandomZoom(
+            height_factor=0.2, width_factor=0.2
+        ),
+    ],
+    name="data_augmentation",
+)
+
+# Now, map the augmentation pipeline to our training dataset
+train_ds_simple = (
+    tf.data.Dataset.from_tensor_slices((x_train, y_train))
+    .shuffle(BATCH_SIZE * 100)
+    .batch(BATCH_SIZE)
+    .map(lambda x, y: (simple_aug(x), y), num_parallel_calls=AUTO)
+    .prefetch(AUTO)
+)
+
+"""
+## Visualize the dataset augmented with RandAugment
+"""
+
+sample_images, _ = next(iter(train_ds_rand))
+plt.figure(figsize=(8, 8))
+for i, image in enumerate(sample_images[:9]):
+    ax = plt.subplot(3, 3, i + 1)
+    plt.imshow(image.numpy().astype("int"))
+    plt.axis("off")
+
+"""
+You are encouraged to run the above code block a couple of times to see different
+variations.
+"""
+
+"""
+## Visualize the dataset augmented with `simple_aug`
+"""
+
+sample_images, _ = next(iter(train_ds_simple))
+plt.figure(figsize=(8, 8))
+for i, image in enumerate(sample_images[:9]):
+    ax = plt.subplot(3, 3, i + 1)
+    plt.imshow(image.numpy().astype("int"))
+    plt.axis("off")
+
+"""
+## Define a model building utility function
+
+Now, we define a shallow CNN. Also, notice that the network already has a rescaling layer
```

Comment:
What do you think about using a bigger model, closer to the state of the art? Maybe a Keras Application model

Reply:
Yes, can do that. Do you have any suggestions in particular? Since we are using the CIFAR-10-C dataset to test out the robustness factor, we need to keep in mind that the models that work better with larger dimensions won't likely be a good fit here.

Reply:
I ran some experiments today with `ResNet50V2`. RandAugment not only just improves the robustness -- from the training logs it seems to have a positive effect on the generalization of the model as well. The training is more stable with it. Here's the [full experimentation notebook](https://nbviewer.jupyter.org/github/sayakpaul/Keras-Examples-RandAugment/blob/main/RandAugment.ipynb).